modules/db/src/main/scala/gem/dao/SmartGcalDao.scala

```diff
@@ -6,34 +6,58 @@ package gem.dao
-
 import doobie.imports._
 
-import scalaz._, Scalaz._
+import scalaz._
+import Scalaz._
+import scalaz.stream.Process
 
 object SmartGcalDao {
 
-  def selectF2(k: SmartGcalKey.F2, t: SmartGcalType): ConnectionIO[List[Int]] =
-    t.fold(Statements.selectF2byLamp(k), Statements.selectF2byBaseline(k)).list
-
-  def select(k: SmartGcalKey, t: SmartGcalType): ConnectionIO[List[GcalConfig]] =
+  def select(k: SmartGcalSearchKey, t: SmartGcalType): ConnectionIO[List[GcalConfig]] =
     for {
       ids <- k match {
-               case f2: SmartGcalKey.F2 => selectF2(f2, t)
+               case f2: SmartGcalKey.F2              => selectF2(f2, t)
+               case gn: SmartGcalKey.GmosNorthSearch => selectGmosNorth(gn, t)
              }
-      gcs <- ids.traverseU { GcalDao.select }.map(_.collect { case Some(a) => a })
+      gcs <- ids.traverseU { GcalDao.selectSmartGcal }.map(_.collect { case Some(a) => a })
     } yield gcs
 
-  def insert(l: GcalLampType, b: GcalBaselineType, k: SmartGcalKey, g: GcalConfig): ConnectionIO[Int] =
-    for {
-      id <- GcalDao.insert(g, None)
-      r  <-
-        k match {
-          case f2: SmartGcalKey.F2 => Statements.insertSmartF2(l, b, id, f2).run
-        }
-    } yield r
+  def selectF2(k: SmartGcalKey.F2, t: SmartGcalType): ConnectionIO[List[Int]] =
+    t.fold(Statements.selectF2ByLamp(k), Statements.selectF2ByBaseline(k)).list
+
+  def selectGmosNorth(k: SmartGcalKey.GmosNorthSearch, t: SmartGcalType): ConnectionIO[List[Int]] =
+    t.fold(Statements.selectGmosNorthByLamp(k), Statements.selectGmosNorthByBaseline(k)).list
+
+  val createIndexF2: ConnectionIO[Int] =
+    Statements.createIndexF2.run
+
+  val dropIndexF2: ConnectionIO[Int] =
+    Statements.dropIndexF2.run
+
+  val createIndexGmosNorth: ConnectionIO[Int] =
+    Statements.createIndexGmosNorth.run
+
+  val dropIndexGmosNorth: ConnectionIO[Int] =
+    Statements.dropIndexGmosNorth.run
+
+  def bulkInsertF2(entries: Vector[(GcalLampType, GcalBaselineType, SmartGcalKey.F2, GcalConfig)]): scalaz.stream.Process[ConnectionIO, Int] =
+    bulkInsert(Statements.bulkInsertF2, entries)
+
+  def bulkInsertGmosNorth(entries: Vector[(GcalLampType, GcalBaselineType, SmartGcalKey.GmosNorthDefinition, GcalConfig)]): scalaz.stream.Process[ConnectionIO, Int] =
+    bulkInsert[SmartGcalKey.GmosNorthDefinition](Statements.bulkInsertGmosNorth, entries)
+
+  private def bulkInsert[K](
+                update:  Update[((GcalLampType, GcalBaselineType, K), Int)],
+                entries: Vector[(GcalLampType, GcalBaselineType, K, GcalConfig)]): scalaz.stream.Process[ConnectionIO, Int] =
+
+    Process.emitAll(entries.map(t => (t._1, t._2, t._3)))        // Process[ConnectionIO, (Lamp, Baseline, Key)]
+      .zip(GcalDao.bulkInsertSmartGcal(entries.map(_._4)))       // Process[ConnectionIO, ((Lamp, Baseline, Key), Id)]
+      .chunk(4096)                                               // Process[ConnectionIO, Vector[((Lamp, Baseline, Key), Id)]]
```

Comment:
Elements. It groups every 4096 `((Lamp, Baseline, Key), Id)` valus into a `Vector`, then we insert each chunk.